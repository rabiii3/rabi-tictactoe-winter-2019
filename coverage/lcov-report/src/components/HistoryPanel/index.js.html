<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/components/HistoryPanel/index.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> / <a href="index.html">src/components/HistoryPanel</a> index.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">70% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>14/20</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/2</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">14.29% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>1/7</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">70% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>14/20</span>
      </div>
    </div>
  </div>
  <div class='status-line medium'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React from 'react';
import PropTypes from 'prop-types';
import { map, reverse, propOr } from 'ramda';
import { Grid, Table, TableBody, TableRow, TableCell, Typography } from '@material-ui/core';
&nbsp;
const HistoryPanel = ({ history }) =&gt; {
  //const round = (history &amp;&amp; history.length + 1) || 0;
  const round = propOr(0, 'length', history);
  return (
    &lt;Grid container direction="column"&gt;
      &lt;Grid item&gt;
        &lt;Status round={round} /&gt;
      &lt;/Grid&gt;
      &lt;Grid item&gt;
        &lt;History history={history} /&gt;
      &lt;/Grid&gt;
    &lt;/Grid&gt;
  );
};
&nbsp;
HistoryPanel.propTypes = {
  history: PropTypes.array.isRequired,
};
&nbsp;
const Status = <span class="fstat-no" title="function not covered" >({</span> round }) =&gt; (
<span class="cstat-no" title="statement not covered" >  &lt;Typography variant="h6" gutterBottom&gt;</span>
    Round: {round}
  &lt;/Typography&gt;
);
&nbsp;
Status.propTypes = {
  round: PropTypes.number.isRequired,
};
&nbsp;
const History = <span class="fstat-no" title="function not covered" >({</span> history }) =&gt; (
<span class="cstat-no" title="statement not covered" >  &lt;Table&gt;</span>
    &lt;TableBody&gt;
      {map(
<span class="fstat-no" title="function not covered" >        ro</span>und =&gt; (
<span class="cstat-no" title="statement not covered" >          &lt;Round round={round} key={round.id} /&gt;</span>
        ),
        reverse(history),
      )}
    &lt;/TableBody&gt;
  &lt;/Table&gt;
);
&nbsp;
History.propTypes = {
  history: PropTypes.array.isRequired,
};
&nbsp;
const Round = <span class="fstat-no" title="function not covered" >({</span> round }) =&gt;
<span class="cstat-no" title="statement not covered" >  round.winner ? &lt;WinnedRound id={round.id} winner={round.winner} /&gt; : &lt;TiedRound id={round.id} /&gt;;</span>
&nbsp;
Round.propTypes = {
  round: PropTypes.object.isRequired,
};
&nbsp;
const WinnedRound = <span class="fstat-no" title="function not covered" >({</span> id, winner }) =&gt; (
<span class="cstat-no" title="statement not covered" >  &lt;TableRow&gt;</span>
    &lt;TableCell&gt;{id}&lt;/TableCell&gt;
    &lt;TableCell&gt;{winner.name}&lt;/TableCell&gt;
  &lt;/TableRow&gt;
);
&nbsp;
WinnedRound.propTypes = {
  id: PropTypes.number.isRequired,
  winner: PropTypes.object.isRequired,
};
&nbsp;
const TiedRound = <span class="fstat-no" title="function not covered" >({</span> id }) =&gt; (
<span class="cstat-no" title="statement not covered" >  &lt;TableRow&gt;</span>
    &lt;TableCell&gt;{id}&lt;/TableCell&gt;
    &lt;TableCell&gt;No Winner&lt;/TableCell&gt;
  &lt;/TableRow&gt;
);
&nbsp;
TiedRound.propTypes = {
  id: PropTypes.number.isRequired,
};
&nbsp;
export default HistoryPanel;
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Mar 14 2019 06:36:09 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
