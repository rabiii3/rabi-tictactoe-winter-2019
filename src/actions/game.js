import { GAME_STARTED, getNewBoard, getNextPlayer, updateBoard, GAME_OVER, isBoardFull, randomPlayer } from '../game';
import { requestComputerPlay, hasAWinner } from '../request';
import { loadFruits } from './fruit';

export const STARTED = 'STARTED';
export const PLAYED = 'PLAYED';
export const PLAYER_CHANGED = 'PLAYER_CHANGED';
export const ENDED = 'ENDED';

export const startGameAction = () => (dispatch, getState) => {
  const {
    game: { player, computer },
  } = getState();
  dispatch({
    type: STARTED,
    payloads: {
      board: getNewBoard(),
      status: GAME_STARTED,
      currentPlayer: randomPlayer(player, computer),
    },
  });
  const {
    game: { currentPlayer },
  } = getState();
  if (currentPlayer.isComputer) dispatch(computerPlay());
};

export const playerPlay = index => (dispatch, getState) => {
  const {
    game: { board, currentPlayer },
  } = getState();
  const newBoard = updateBoard(board, index, currentPlayer);
  dispatch(manageGame(newBoard));
};

const computerPlay = () => (dispatch, getState) => {
  const {
    game: { board, currentPlayer, player, computer },
  } = getState();
  return requestComputerPlay({ board, computer: computer.piece, player: player.piece }).then(({ move }) => {
    const newBoard = updateBoard(board, move, currentPlayer);
    setTimeout(() => dispatch(manageGame(newBoard)), 1000);
  });
};

export const manageGame = newBoard => (dispatch, getState) => {
  dispatch({
    type: PLAYED,
    payloads: {
      board: newBoard,
    },
  });

  const {
    game: { board, currentPlayer, player, computer },
  } = getState();
  hasAWinner({ board }).then(({ winner }) => {
    if (winner || isBoardFull(board)) {
      dispatch(loadFruits()).then(() =>
        dispatch({
          type: ENDED,
          payloads: {
            result: winner,
            player,
            computer,
            status: GAME_OVER,
          },
        }),
      );
    } else {
      dispatch({
        type: PLAYER_CHANGED,
        payloads: {
          currentPlayer: getNextPlayer(currentPlayer, player, computer),
        },
      });

      const {
        game: { currentPlayer: cp },
      } = getState();
      console.log('======currennt player :', cp);
      if (cp.isComputer) dispatch(computerPlay());
    }
  });
};
