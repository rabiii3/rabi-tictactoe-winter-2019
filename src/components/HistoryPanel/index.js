import React from 'react';
import PropTypes from 'prop-types';
import { map, head } from 'ramda';
import { pure } from 'recompose';
import { Grid, Table, TableBody, TableRow, TableCell, Typography, withStyles } from '@material-ui/core';

const styles = {
  container: {
    width: 350,
    height: 350,
  },
};

const HistoryPanel = ({ history, classes }) => {
  //const round = propOr(0, 'length', history);
  return (
    <Grid container direction="column" className={classes.container}>
      <Grid item>
        <Status round={head(history)} />
      </Grid>
      <Grid item>
        <History history={history} />
      </Grid>
    </Grid>
  );
};

HistoryPanel.propTypes = {
  history: PropTypes.array.isRequired,
  classes: PropTypes.object.isRequired,
};

const PureHistoryPanel = pure(HistoryPanel);

const Status = ({ round }) => (
  <Typography variant="h6" gutterBottom align="center">
    Round: {round.id + 1}
  </Typography>
);

Status.propTypes = {
  round: PropTypes.object.isRequired,
};

const History = ({ history }) => (
  <Table>
    <TableBody>
      {map(
        round => (
          <Round round={round} key={round.id} />
        ),
        history,
      )}
    </TableBody>
  </Table>
);

History.propTypes = {
  history: PropTypes.array.isRequired,
};

const Round = ({ round }) =>
  round.winner ? <WinnedRound id={round.id} winner={round.winner} /> : <TiedRound id={round.id} />;

Round.propTypes = {
  round: PropTypes.object.isRequired,
};

const WinnedRound = ({ id, winner }) => (
  <TableRow>
    <TableCell>{id}</TableCell>
    <TableCell>{winner.name}</TableCell>
  </TableRow>
);

WinnedRound.propTypes = {
  id: PropTypes.number.isRequired,
  winner: PropTypes.object.isRequired,
};

const TiedRound = ({ id }) => (
  <TableRow>
    <TableCell>{id}</TableCell>
    <TableCell>No Winner</TableCell>
  </TableRow>
);

TiedRound.propTypes = {
  id: PropTypes.number.isRequired,
};

export default withStyles(styles)(PureHistoryPanel);
