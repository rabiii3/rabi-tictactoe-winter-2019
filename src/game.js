import { indexOf, update, propOr } from 'ramda';

export const X = 'x';

export const O = 'o';

export const GAME_OVER = 'GAME_OVER';

export const GAME_STARTED = 'GAME_STARTED';

export const isBoardFull = board => indexOf(null, board) < 0;

export const getStatus = board => (isBoardFull(board) ? GAME_OVER : GAME_STARTED);

export const getNewComputerBoard = (board, computer) => update(indexOf(null, board), computer.piece, board);

export const getNewBoard = () => [null, null, null, null, null, null, null, null, null];

export const getFirstAvailableCell = board => indexOf(null, board);

export const updateBoard = (board, index, currentPlayer) =>
  isBoardFull(board) ? board : update(index, currentPlayer.piece, board);

export const getNextPlayer = (currentPlayer, player, computer) => (currentPlayer.isComputer ? player : computer);

const random = () => Math.round(Math.random());

export const randomPlayer = (player, computer) => {
  const players = [player, computer];
  return players[random()];
};

export const updateFruitBoard = (board, index, icon) => update(index, icon, board);

export const isFruit = piece => (piece != null && propOr(false, 'icon', piece) ? true : false);
