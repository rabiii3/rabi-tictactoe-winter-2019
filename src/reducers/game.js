import { STARTED, PLAYED, PLAYER_CHANGED, ENDED } from '../actions/game';
import { CELL_FRUIT_LOADED } from '../actions/fruit';
import { updateFruitBoard } from '../game';

export default (state = {}, action) => {
  var i = 0;
  switch (action.type) {
    case STARTED:
      return { ...state, ...action.payloads };
    case PLAYED:
      return { ...state, ...action.payloads };
    case PLAYER_CHANGED:
      return { ...state, ...action.payloads };
    case ENDED: {
      const { status } = action.payloads;
      return { ...state, status };
    }
    case CELL_FRUIT_LOADED: {
      const { board } = state;
      const { index, fruit } = action;
      return { ...state, board: updateFruitBoard(board, index, fruit) };
    }
    default:
      return state;
  }
};
