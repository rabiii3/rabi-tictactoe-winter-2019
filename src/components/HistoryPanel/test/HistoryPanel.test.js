import React from 'react';
import { shallow } from 'enzyme';
import { X, O } from '../../../game';
import HistoryPanel from '..';

describe('HistoryPanel', () => {
  it('should render', () => {
    const player = { name: 'player', isComputer: false, piece: O };
    const computer = { name: 'computer', isComputer: true, piece: X };
    const history = [{ id: 1, winner: player }, { id: 2, winner: computer }, { id: 3, winner: player }, { id: 4 }];
    const root = <HistoryPanel history={history} />;
    const wrapped = shallow(root);
    expect(wrapped).toMatchSnapshot();
  });
});
