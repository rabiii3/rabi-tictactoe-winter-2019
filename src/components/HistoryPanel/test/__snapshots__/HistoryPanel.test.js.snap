// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryPanel should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HistoryPanel
    history={
      Array [
        Object {
          "id": 1,
          "winner": Object {
            "isComputer": false,
            "name": "player",
            "piece": "o",
          },
        },
        Object {
          "id": 2,
          "winner": Object {
            "isComputer": true,
            "name": "computer",
            "piece": "x",
          },
        },
        Object {
          "id": 3,
          "winner": Object {
            "isComputer": false,
            "name": "player",
            "piece": "o",
          },
        },
        Object {
          "id": 4,
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <WithStyles(Grid)
          item={true}
        >
          <Status
            round={4}
          />
        </WithStyles(Grid)>,
        <WithStyles(Grid)
          item={true}
        >
          <History
            history={
              Array [
                Object {
                  "id": 1,
                  "winner": Object {
                    "isComputer": false,
                    "name": "player",
                    "piece": "o",
                  },
                },
                Object {
                  "id": 2,
                  "winner": Object {
                    "isComputer": true,
                    "name": "computer",
                    "piece": "x",
                  },
                },
                Object {
                  "id": 3,
                  "winner": Object {
                    "isComputer": false,
                    "name": "player",
                    "piece": "o",
                  },
                },
                Object {
                  "id": 4,
                },
              ]
            }
          />
        </WithStyles(Grid)>,
      ],
      "container": true,
      "direction": "column",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Status
            round={4}
          />,
          "item": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "round": 4,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <History
            history={
              Array [
                Object {
                  "id": 1,
                  "winner": Object {
                    "isComputer": false,
                    "name": "player",
                    "piece": "o",
                  },
                },
                Object {
                  "id": 2,
                  "winner": Object {
                    "isComputer": true,
                    "name": "computer",
                    "piece": "x",
                  },
                },
                Object {
                  "id": 3,
                  "winner": Object {
                    "isComputer": false,
                    "name": "player",
                    "piece": "o",
                  },
                },
                Object {
                  "id": 4,
                },
              ]
            }
          />,
          "item": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "history": Array [
              Object {
                "id": 1,
                "winner": Object {
                  "isComputer": false,
                  "name": "player",
                  "piece": "o",
                },
              },
              Object {
                "id": 2,
                "winner": Object {
                  "isComputer": true,
                  "name": "computer",
                  "piece": "x",
                },
              },
              Object {
                "id": 3,
                "winner": Object {
                  "isComputer": false,
                  "name": "player",
                  "piece": "o",
                },
              },
              Object {
                "id": 4,
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(Grid)
            item={true}
          >
            <Status
              round={4}
            />
          </WithStyles(Grid)>,
          <WithStyles(Grid)
            item={true}
          >
            <History
              history={
                Array [
                  Object {
                    "id": 1,
                    "winner": Object {
                      "isComputer": false,
                      "name": "player",
                      "piece": "o",
                    },
                  },
                  Object {
                    "id": 2,
                    "winner": Object {
                      "isComputer": true,
                      "name": "computer",
                      "piece": "x",
                    },
                  },
                  Object {
                    "id": 3,
                    "winner": Object {
                      "isComputer": false,
                      "name": "player",
                      "piece": "o",
                    },
                  },
                  Object {
                    "id": 4,
                  },
                ]
              }
            />
          </WithStyles(Grid)>,
        ],
        "container": true,
        "direction": "column",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Status
              round={4}
            />,
            "item": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "round": 4,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <History
              history={
                Array [
                  Object {
                    "id": 1,
                    "winner": Object {
                      "isComputer": false,
                      "name": "player",
                      "piece": "o",
                    },
                  },
                  Object {
                    "id": 2,
                    "winner": Object {
                      "isComputer": true,
                      "name": "computer",
                      "piece": "x",
                    },
                  },
                  Object {
                    "id": 3,
                    "winner": Object {
                      "isComputer": false,
                      "name": "player",
                      "piece": "o",
                    },
                  },
                  Object {
                    "id": 4,
                  },
                ]
              }
            />,
            "item": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "history": Array [
                Object {
                  "id": 1,
                  "winner": Object {
                    "isComputer": false,
                    "name": "player",
                    "piece": "o",
                  },
                },
                Object {
                  "id": 2,
                  "winner": Object {
                    "isComputer": true,
                    "name": "computer",
                    "piece": "x",
                  },
                },
                Object {
                  "id": 3,
                  "winner": Object {
                    "isComputer": false,
                    "name": "player",
                    "piece": "o",
                  },
                },
                Object {
                  "id": 4,
                },
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
